crates/ide/src/completion/complete_postfix/format_like.rs

```diff
@@ -0,0 +1,325 @@
+//! Postfix completion for `format`-like strings.
+//!
+//! `"Result {result} is {2 + 2}"` is expanded to the `"Result {} is {}", result, 2 + 2`.
+//!
+//! The following postfix snippets are available:
+//!
+//! - `format` -> `format!(...)`
+//! - `panic` -> `panic!(...)`
+//! - `println` -> `println!(...)`
+//! - `log`:
+//!   + `logd` -> `log::debug!(...)`
+//!   + `logt` -> `log::trace!(...)`
+//!   + `logi` -> `log::info!(...)`
+//!   + `logw` -> `log::warn!(...)`
+//!   + `loge` -> `log::error!(...)`
+
+use super::postfix_snippet;
+use crate::completion::{
+    completion_config::SnippetCap, completion_context::CompletionContext,
+    completion_item::Completions,
+};
+use syntax::ast;
+
+pub(super) fn add_format_like_completions(
+    acc: &mut Completions,
+    ctx: &CompletionContext,
+    dot_receiver: &ast::Expr,
+    cap: SnippetCap,
+    receiver_text: &str,
+) {
+    if !is_string_literal(receiver_text) {
+        // It's not a string literal, do not parse input.
+        return;
+    }
+
+    let input = &receiver_text[1..receiver_text.len() - 1];
+
+    let mut parser = FormatStrParser::new(input);
+
+    if parser.parse().is_ok() {
+        for kind in PostfixKind::all_suggestions() {
+            let snippet = parser.into_suggestion(*kind);
+            let (label, detail) = kind.into_description();
+
+            postfix_snippet(ctx, cap, &dot_receiver, label, detail, &snippet).add_to(acc);
+        }
+    }
+}
+
+/// Checks whether provided item is a string literal.
+fn is_string_literal(item: &str) -> bool {
+    if item.len() < 2 {
+        return false;
+    }
+    item.starts_with("\"") && item.ends_with("\"")
+}
```

Comment:
Just as a general code-style issue, the call to `is_string` and subsequent `&receiver_text[1..receiver_text.len() - 1];` indexing have an invariant (`1` and `-1` are utf8 boundaries) spanning a function call. It's best to avoid such inter-function invariants by making `is_string_literal` return an `Option<&str>`.